Xchess/xchess.1
.TH XCHESS 1 "14 Nov 1986" "X Version 10"
.SH NAME
xchess \- X chess display
.SH SYNOPSIS
.B xchess
[ option ... ] [ white-display ] [ black-display ]
.SH DESCRIPTION
.PP
.B xchess
is a chess display program which allows players to play a game on either
one or two displays, or play a chess-playing program.  It uses the
.B X
window system.  If one or no display names are given, it will open up one
window and both black and white at the same board.  If two displays are
given,
.B xchess
will accept moves from each player in his turn.  Black's board will be drawn
with his pieces at the bottom.
.PP
.B xchess
will not allow a player to make an illegal move.  It accepts all legal moves,
including castling and pawn capture \fIen passant\fR.
.SH OPTIONS
.TP 8
.B -d
Turn on debugging.
.TP 8
.B -f record-file
Use \fBrecord-file\fR for saving the game when the \fBSave\fR button is
selected, or if the \fB-s\fR flag is given.  The default is "xchess.game".
.TP 8
.B -r saved-game
Start with the position at the end of the saved game in the named file.
This file may be the result of the \fBSave\fR command, and may be in
either English or International format.  When reading moves, one move
it made per second.
.TP 8
.B -q
Don't pause for a second every time a move is made when a game is being
restored.
.TP 8
.B -v
Whenever a piece is moved, outline the path with a "lightning bolt".
This option and the \fB-n\fR option are useful if you don't want to miss
an opponent's move when he makes it.
.TP 8
.B -i
Use International format for recording moves (squares numbered 1-8, a-h)
as opposed to English (e.g, \fIp/k4xp/q5\fR).
.TP 8
.B -t moves/timeunit
Allows \fBtimeunit\fR seconds for every \fBmoves\fR moves.  If either player
exceeds this allowance both recieve a message saying informing them of
this fact.
.TP 8
.B -c
Play the computer.
.B xchess
will start up a chess-playing program (currently the only one it knows
how to talk to is \fBGNU Chess\fR).
.TP 8
.B -p program
The name of the program to use if the \fB-c\fR option is given.  The
default is "/usr/public/gnuchess".  Note that \fBgnuchess\fR must be
compiled with the \fIcompat\fR flag (in the file "main.c") set to 1.
.TP 8
.B -b
If the \fB-c\fR flag was given, have the computer play white.
.TP 8
.B -bnw
If the display has more than one display plane (i.e, is color), pretend
it's black and white.
.TP 8
.B -s
Save the moves in the record file as they are made.  This is useful if
you don't want your game to be lost when \fBxchess\fR core dumps.
.TP 8
.B -n
Be noisy \- beep after every move is made.
.TP 8
.B -h host
Run GNU Chess on the specified \fBhost\fR.
.TP 8
.B -R
Randomly chose who plays white and who plays black, if two displays are
given.
.SH CONTROLS
.PP
The window is divided up into several sub-windows.  The pieces are moved by
pushing down any mouse button on top of the piece, moving to the destination
square, and releasing it.  Castling is done by moving the king to the
right square.  If you push down on a piece and then let the button
up without moving it, you must move that piece. ("Touch it, move it.")
.PP
The progress of the game is listed in the "Game Record" window.  Error
messages and such things are printed in the "Message" window.  Both these
windows have scroll bars that you can use to move around.
There are also windows for clocks and for a record of the pieces captured.
.PP
If you type any keys in the window, the text will go into the message
window of both players.  This provides a simple communication facility.
.PP
There are 9 buttons in the control window.  They are as follows:
.TP 8
.B Draw
Both players must push this button to agree on a draw (just one is ok
if only one display is being used).
.TP 8
.B Resign
The player whose turn it is to move resigns.
.TP 8
.B Reset
Start over from the beginning.
.TP 8
.B Back
Retract a move.  If two displays are being used the other player will be
asked to confirm this.
.TP 8
.B Fwd
This will re-play the most recently retracted move.  This button in conjunction
with \fBBack\fR is useful for "scrolling around" in a saved game.
.TP 8
.B Save
Save the game in the record file.
.TP 8
.B Flip
Rotate the board so that Black will have his pieces at the bottom.
.TP 8
.B Switch
Change the mapping of boards to players.
.TP 8
.B Pause
This button has two functions.  When a game is being restored, pieces will
be moved once a second.  Hitting \fBPause\fR will stop this process, and
hitting it again will restart it.  During the time that it is stopped no
other action can be made except restarting it.  While a game is being played,
\fBPause\fR will stop the clock and restart it.
.SH DEFAULTS
.PP
\fBxchess\fR uses the following \fI.Xdefaults\fR:
.TP 8
.B Noisy
The -n flag.
.TP 8
.B SaveMoves
The -s flag.
.TP 8
.B Algebraic
The -i flag.
.TP 8
.B BlackAndWhite
The -bnw flag.
.TP 8
.B QuickRestore
The -q flag.
.TP 8
.B Flash
The -v flag.
.TP 8
.B NumFlashes
How many times to flash the move.  The default is 5.
.TP 8
.B FlashWidth
How big to make the lightning bolt.  The default is 10 pixels.
.TP 8
.B ProgName
The -p option.  This may also be changed in the Makefile (-DDEF_PROG_NAME).
.TP 8
.B ProgHost
The -h option.
.TP 8
.B RecordFile
The -f option.
.TP 8
.B BlackPiece
The color of the black pieces.
.TP 8
.B WhitePiece
The color of the white pieces.
.TP 8
.B BorderColor
The color of the borders.
.TP 8
.B BlackSquare
The color of the black squares.
.TP 8
.B WhiteSquare
The color of the white squares.
.TP 8
.B TextColor
The color of routine messages and the move record text.
.TP 8
.B ErrorText
The color of error messages.
.TP 8
.B PlayerText
The color of player-entered text.
.TP 8
.B TextBack
The background color for the two text windows.
.TP 8
.B CursorColor
The color of the mouse and the text cursors.
.SH "SEE ALSO"
X(8), gnuchess(1), chess(5)
.SH AUTHOR
Wayne A. Christopher (faustus@ic.berkeley.edu)
.SH BUGS
.PP
Checkmate and stalemate are not detected, so the appropriate player must resign
or agree to a draw respectively.
.PP
\fBSwitch\fR doesn't work.
.PP
If you are playing \fBgnuchess\fR, and you select Undo a few times so that it
is \fBgnuchess\fR's turn to move, it won't do anything.


CHANGES
			  GNU CHESS HISTORY
		      (current version 1.2.87)


January 2, 1987   -- Stuart Cracraft at UCLA
		     He added a few more Tal games to the collection.

January 2, 1987   -- Jim Aspnes at MIT
		     He contributed MCO variations for the Catalan,
		     Queen's Indian, and Reti openings.

December 29, 1986 -- Jim Aspnes at MIT
		     He contributed all MCO variations of the Najdorf
		     to the opening book.He also contributed a LISP
		     macro (written in GNU Emacs Lisp) to convert
		     xchess game formats to GNU Chess opening book
		     format. 

December 14, 1986 -- Ken Thompson at Bell Labs
		     He contributed almost 200 games by Tal to
		     our collection of Tal-games, bringing the
		     total number of Tal positions in the book
		     to 10,692. Total book positions now 13,207.
		     These reside in bookin, bookin.bdg, bookin.tal.
		     Note that presently, only bookin and bookin.tal
		     can be used. The new Tal positions came in a
		     slightly different format, which we have chosen
		     to adopt as our standard format. All book
		     games in bookin and bookin.bdg will gradually
		     change into the new standard format.

December 11, 1986 -- Stuart Cracraft at UCLA
		     Added "averages" for node-count per move,
		     cpu per move, rate per move to list_history
		     and write_history.
		     New version of Xchess installed.
		     Started typing in Tal games into "bookin.tal".
		     Added "total book positions" printout to "book"
		     and "enter" statistics printout.

December 10, 1986 -- Stuart Cracraft at UCLA
		     Implemented aspiration search in normal
		     alpha-beta search. Speedups of 3% to 40%
		     have been noticed in most positions.
		     Occasionally a slower search will result,
		     but it is thought these are worth the
		     usual speedups.

December 9, 1986  -- Stuart Cracraft at UCLA
		     Fixed minor bug in write_history()
		     Added another Tal game, 2nd game of 1st world
		     championship match with Botvinnik, a Benoni.

December 9, 1986  -- Stuart Cracraft at UCLA
		     Enhanced parallelism. All parallel processors
		     now communicate via a shared data file and
		     are kept running (in idle loops watching the
		     shared data file). This saves us a few seconds
		     on each move since the 'rsh' need not be invoked
		     more than once (at the beginning). Since the
		     shared data file is now implemented, we will
		     next work towards a "parallel anarchy" in which
		     any processor can use any other processor in
		     order to reduce its search. The current scheme
		     with the program being only as fast as its slowest
		     processor, is quite inefficient.

December 1, 1986 --  ASP at MIT Athena
		     Added a couple of Master games from 
		     Modern Chess Openings 12 (a Fischer game,
		     and a Matanovic game).

November 30, 1986 -- Stuart Cracraft at UCLA
		     Added parallelism. Can now handle multiple
		     processors (sharing same disk). Later we will
		     add the capability to use processors not sharing
		     the same disk. Modified README and MAN-PAGE.

November 26, 1986 -- Stuart Cracraft at UCLA
		     Fixed a few bugs in book-mailing mechanism.
		     Fixed a bug regarding situations where only
		     one move is available.
		     Fixed a bug in read_history() that caused
		     Black queenside castles to be mishandled.

November 25, 1986 -- Stuart Cracraft at UCLA
		     Added two pawn heuristics. Reward pawns moving into
		     a phalanx of pawns. A phalanx is two or more
		     horizontally-connected pawns. Likewise, penalize
		     pawns leaving a phalanx of pawns. The penalty for
		     leaving is a little more than the reward for
		     entering.

November 24, 1986 -- Stuart Cracraft at UCLA
		     A user reported an unbelievable bug. Investigation
		     of this bug led to the discovery that GNU Chess was
 		     not picking the move judged best by the tree search
		     in all cases. This resulted in the bug showing
 		     itself which further showed that the program was
		     selecting an inferior move. This may result in an
		     improvement to the program's play.

November 24, 1986 -- Stuart Cracraft at UCLA
		     Added two heuristics. Penalize king moves if
		     the king hasn't castled yet. Also, penalize pawn
		     moves which produce doubled pawns. Should
		     probably have something for isolated pawns
		     too.

November 23, 1986 -- Wayne Christopher at Berkeley
		     New version of X chess display front-end.
		     Fixed bugs include multiple pieces, runs
		     on SUNS & Bobcats, loads saved games.

November 23, 1986 -- Stuart Cracraft at UCLA
		     Cleaned up some minor bugs regarding history.
		     Added "Illegal command" error message at Wayne's
		     request.

November 22, 1986 -- David Goldberg at SUN Microsystems
		     He complained that GNU Chess was memory-hungry.
		     A few minor modifications to hash.c reduced
		     uninitialized data space 87% and text space
		     12%. This should make it easier for GNU Chess
		     to run on small computers.

November 22, 1986 -- Stuart Cracraft at UCLA
		     "read" command was working, but needed
		     additional tweaking so that history
		     array would be printed by list_history().

November 19, 1986 -- Stuart Cracraft at UCLA
		     Added "read" command which reads a history
		     file (game listing) and restores the board
		     to as if the person was still playing that.
		     particular game. Generally cleaned up 
		     history mechanism, made it more orthogonal.
		     Revised README. Added doc to MAN-PAGE.

November 16, 1986 -- Stuart Cracraft at UCLA
		     More opening book bugs found and fixed.
		     Added capability to accept abbreviated-algebraic notation
		     for entering "book" games from files.
		     Added approximately 2500 new positions to
		     opening book from games involving the
		     opening called Blackmar-Diemer Gambit,

Xchess/xchess.c
/* This file contains code for X-CHESS.
   Copyright (C) 1986 Free Software Foundation, Inc.

This file is part of X-CHESS.

X-CHESS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY.  No author or distributor
accepts responsibility to anyone for the consequences of using it
or for whether it serves any particular purpose or works at all,
unless he says so in writing.  Refer to the X-CHESS General Public
License for full details.

Everyone is granted permission to copy, modify and redistribute
X-CHESS, but only under the conditions described in the
X-CHESS General Public License.   A copy of this license is
supposed to have been given to you along with X-CHESS so you
can know your rights and responsibilities.  It should be in a
file named COPYING.  Among other things, the copyright notice
and this notice must be preserved on all copies.  */


/* RCS Info: $Revision: 1.5 $ on $Date: 86/11/26 12:11:32 $
 *           $Source: /users/faustus/xchess/RCS/xchess.c,v $
 * Copyright (c) 1986 Wayne A. Christopher, U. C. Berkeley CAD Group
 *	Permission is granted to do anything with this code except sell it
 *	or remove this message.
 */

#define USAGE	"xchess [ -d ] [ -f recordfile ] [ -r savedfile ] [ -i ]\n\
\t[ -t moves/timeunit ] [ -c ] [ -p program ]  [ -b ] [ -bnw ] [ -s ]\n\
\t[ -n ] [ -h host ] [ -v ] [ -R ] [ whitedisplay ] [ blackdisplay ]"

#include "xchess.h"

bool debug = false;
bool oneboard = false;
bool bnwflag = false;
bool progflag = false;
bool blackflag = false;
bool quickflag = false;

char *progname = DEF_PROGRAM;
char *proghost = NULL;
char *piecenames[] = { "pawn", "rook", "knight", "bishop", "queen", "king" } ;
char *colornames[] = { "white", "black", "none" } ;
char *movetypenames[] = { "move", "qcastle", "kcastle", "capture" } ;
char *dispname1 = NULL, *dispname2 = NULL;

char *black_piece_color = BLACK_PIECE_COLOR;
char *white_piece_color = WHITE_PIECE_COLOR;
char *black_square_color = BLACK_SQUARE_COLOR;
char *white_square_color = WHITE_SQUARE_COLOR;
char *border_color = BORDER_COLOR;
char *text_color = TEXT_COLOR;
char *text_back = TEXT_BACK;
char *error_text = ERROR_TEXT;
char *player_text = PLAYER_TEXT;
char *cursor_color = CURSOR_COLOR;

int num_flashes = NUM_FLASHES;
int flash_size = FLASH_SIZE;

void
main(ac, av)
	char **av;
{
	char *program = av[0];
	char *recfile = NULL, *s;
	bool randflag = false;
	move *m;

	/* Process args. */
	av++; ac--;
	while (**av == '-') {
		if (eq(*av, "-d")) {
			debug = true;
		} else if (eq(*av, "-f")) {
			av++; ac--;
			if (*av)
				record_file = *av;
			else
				goto usage;
		} else if (eq(*av, "-r")) {
			av++; ac--;
			if (*av)
				recfile = *av;
			else
				goto usage;
		} else if (eq(*av, "-i")) {
			record_english = false;
		} else if (eq(*av, "-R")) {
			randflag = true;
		} else if (eq(*av, "-v")) {
			win_flashmove = true;
		} else if (eq(*av, "-q")) {
			quickflag = true;
		} else if (eq(*av, "-t")) {
			av++; ac--;
			if (*av) {
				movesperunit = atoi(*av);
				if (s = index(*av, '/'))
					timeunit = atoi(s + 1);
				else
					timeunit = 60;
			} else
				goto usage;
		} else if (eq(*av, "-p")) {
			av++; ac--;
			if (*av)
				progname = *av;
			else
				goto usage;
		} else if (eq(*av, "-h")) {
			av++; ac--;
			if (*av)
				proghost = *av;
			else
				goto usage;
		} else if (eq(*av, "-b")) {
			blackflag = true;
		} else if (eq(*av, "-c")) {
			progflag = true;
		} else if (eq(*av, "-bnw")) {
			bnwflag = true;
		} else if (eq(*av, "-s")) {
			saveflag = true;
		} else if (eq(*av, "-n")) {
			noisyflag = true;
		} else
			goto usage;
		av++; ac--;
	}
	if (ac > 0)
		dispname1 = av[0];
	if (ac > 1)
		dispname2 = av[1];
	if (ac > 2)
		goto usage;

	if (!dispname2)
		oneboard = true;
	
	srandom(getpid());

	if (!oneboard && randflag && (random() % 2)) {
		s = dispname1;
		dispname1 = dispname2;
		dispname2 = s;
	}
		
	if (!dispname1)
		dispname1 = getenv("DISPLAY");
	
	/* Now get boolean defaults... */
	if ((s = XGetDefault(program, "noisy")) && eq(s, "on"))
		noisyflag = true;
	if ((s = XGetDefault(program, "savemoves")) && eq(s, "on"))
		saveflag = true;
	if ((s = XGetDefault(program, "algebraic")) && eq(s, "on"))
		record_english = false;
	if ((s = XGetDefault(program, "blackandwhite")) && eq(s, "on"))
		bnwflag = true;
	if ((s = XGetDefault(program, "quickrestore")) && eq(s, "on"))
		quickflag = true;
	if ((s = XGetDefault(program, "flash")) && eq(s, "on"))
		win_flashmove = true;
	
	/* ... numeric variables ... */
	if (s = XGetDefault(program, "numflashes"))
		num_flashes = atoi(s);
	if (s = XGetDefault(program, "flashsize"))
		flash_size = atoi(s);
	
	/* ... and strings. */
	if (s = XGetDefault(program, "progname"))
		progname = s;
	if (s = XGetDefault(program, "proghost"))
		proghost = s;
	if (s = XGetDefault(program, "recordfile"))
		recfile = s;
	if (s = XGetDefault(program, "blackpiece"))
		black_piece_color = s;
	if (s = XGetDefault(program, "whitepiece"))
		white_piece_color = s;
	if (s = XGetDefault(program, "blacksquare"))
		black_square_color = s;
	if (s = XGetDefault(program, "whitesquare"))
		white_square_color = s;
	if (s = XGetDefault(program, "bordercolor"))
		border_color = s;
	if (s = XGetDefault(program, "textcolor"))
		text_color = s;
	if (s = XGetDefault(program, "textback"))
		text_back = s;
	if (s = XGetDefault(program, "errortext"))
		error_text = s;
	if (s = XGetDefault(program, "playertext"))
		player_text = s;
	if (s = XGetDefault(program, "cursorcolor"))
		cursor_color = s;

	/* Set up the board. */
	board_setup();

	/* Create the windows. */
	win_setup(dispname1, dispname2);

	board_drawall();

	/* Start the program if necessary. */
	if (progflag)
		if (!program_init(progname))
			exit(1);
	
	XSync(0);
	if (recfile)
		load_game(recfile);

	/* Go into a loop of prompting players alternately for moves, checking
	 * them, and updating things.
	 */
	for (;;) {
		win_process(false);
		clock_update();
		if (progflag && ((!blackflag && (nexttomove == BLACK)) ||
				(blackflag && (nexttomove == WHITE)))) {
			m = program_get();
			if (m)
				prog_move(m);
		}
	}

usage:	fprintf(stderr, "Usage: %s\n", USAGE);
	exit(1);
}

